### ISSUE TITLE
follow-up actions for array API support in `cluster`

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#18866]
These are follow-up items from gh-18668:

- [x] a better name for `as_xparray` (see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1631049650))
- [x] a TODO comment in `hierarchy.py` for rewriting a fancy indexing expression (may need `xp.put` to be standardized, see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1594618749))
- [x] there's a usage of `xp.cov` in `hierarchy.py` that needs changing (it's in numpy/cupy/pytorch, but not in the standard)
- [x] there's a usage of `xp.searchsorted` in `hierarchy.py` that either needs changing (it's in numpy/cupy/pytorch, but not in the standard) or the standardization for it needs to be pushed forward (it does make sense to include in the standard probably).
- [x] consider removing the `skip_if_array_api` marker (see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1614945157) and gh-19181)
- [x] (after we drop `setup.py`) making the import of the vendored `array_api_compat` nicer (see [this comment](https://github.com/scipy/scipy/pull/18668#issuecomment-1607878468))
- [x] behavior of _array-like_'s ~~(see gh-19118)~~(array-likes are now allowed, but fine tuning is needed. See https://github.com/scipy/scipy/pull/19187#issuecomment-1708877650)

**Comment by rgommers** [comment]
[Comment from @asmeurer](https://github.com/scipy/scipy/pull/18668#issuecomment-1634721495) about the `as_xparray` name follow-up, copied here:

_I think we decided that this specific wrapper was doing things that weren't appropriate for general inclusion in the compat library (c.f. https://github.com/data-apis/array-api-compat/pull/42), but an option could be to monkeypatch `array_api_compat.asarray`, especially since you are vendoring it._

I suspect a regular good name is preferable over monkey-patching (clearly separating what's standard from what's custom, which helps not only for clarity but will be useful to figure out if/when to drop compat layer helpers at a point when the standard is fully supported in array libraries).

No one has suggested a new name yet though. To trigger some ideas, here is the current signature
```python
def as_xparray(
    array, dtype=None, order=None, copy=None, *, xp=None, check_finite=False
):
```
and also the signature of what is perhaps its closest equivalent, [_asarray_validated](https://github.com/scipy/scipy/blob/28eecdb22ec5e3e621dd515bccc158f2f7b25b4c/scipy/_lib/_util.py#L201):
```python
def _asarray_validated(a, check_finite=True,
                       sparse_ok=False, objects_ok=False, mask_ok=False,
                       as_inexact=False):
```

What `as_xparray` has extra compared to `xp.asarray` is:
- the `order` and `copy` keywords, which can be used if and only if the input is a numpy array
- the `check_finite` keyword, which (if set to True) makes the function raise a `ValueError` if `inf`/`nan` values are present
- if `xp` is not given as a keyword, it also calls the internal `compliance_scipy` utility which raises for `np.matrix`, masked arrays and object arrays.

**Comment by asmeurer** [comment]
Monkeypatch or no, my suggestion would still be to just call it `asarray` (or `_asarray` if it really bothers you that it's a different signature than the usual one).

**Comment by lucascolley** [comment]
Like `xp.cov`, `xp.searchsorted` is also not in the standard (originally noted by Ralf here https://github.com/scipy/scipy/pull/18668/files#r1228672460):
https://github.com/scipy/scipy/blob/604e7a28e145dd14b77666752b65879bb73b01d7/scipy/cluster/hierarchy.py#L1343-L1346

**Comment by lucascolley** [comment]
> a TODO comment in hierarchy.py for rewriting a fancy indexing expression

For future reference, once this is done, the `@skip_if_array_api_backend('numpy.array_api')` here can be removed :
https://github.com/scipy/scipy/blob/604e7a28e145dd14b77666752b65879bb73b01d7/scipy/cluster/tests/test_hierarchy.py#L1200-L1201

**Comment by lucascolley** [comment]
The `xp.cov` task can be checked off now that #19051 is merged, and replaced with `xp.searchsorted` as mentioned above.

**Comment by rgommers** [comment]
Done. Added a note that `searchsorted` may be standardized. So not urgent to try and replace it.

**Comment by lucascolley** [comment]
> behavior of array-like's (see https://github.com/scipy/scipy/issues/19118)(array-likes are now allowed, but fine tuning is needed. See https://github.com/scipy/scipy/pull/19187#issuecomment-1708877650)

The relevant discussion for this point is collected in gh-19276.

**Comment by lucascolley** [comment]
> there's a usage of xp.searchsorted in hierarchy.py that either needs changing (it's in numpy/cupy/pytorch, but not in the standard) or the standardization for it needs to be pushed forward (it does make sense to include in the standard probably).

Good idea to link to https://github.com/data-apis/array-api/issues/688

**Comment by lucascolley** [comment]
> behavior of array-like's (see BUG: np-coercible array-likes are not accepted with array API flag set #19118)(array-likes are now allowed, but fine tuning is needed. See BUG: accept np-coercible array-likes with array API flag set #19187 (comment))

This can be ticked off now following gh-19606.

**Comment by lucascolley** [comment]
> (after we drop setup.py) making the import of the vendored array_api_compat nicer (see https://github.com/scipy/scipy/pull/18668#issuecomment-1607878468)

completed by gh-19701

**Comment by rgommers** [comment]
> there's a usage of `xp.searchsorted` in `hierarchy.py` that either needs changing (it's in numpy/cupy/pytorch, but not in the standard) or the standardization for it needs to be pushed forward (it does make sense to include in the standard probably)

https://github.com/data-apis/array-api/pull/699 should add `searchsorted` to the standard, so there's nothing else to do on our side right now.

**Comment by lucascolley** [comment]
the `searchsorted` point can be ticked off now (however I don't think we can remove any test skips yet - `cut_tree` still has a `# TODO ARRAY_API complex indexing not supported`

**Comment by lucascolley** [comment]
> consider removing the skip_if_array_api marker

completed ✅

**Comment by rgommers** [comment]
All the original follow-up items were addressed. So much has been changed/improved in `cluster` that this issue is useful anymore, even if there's still one TODO comment floating around. We can track the state of things better via `xp_capabilities` and the skips in the test suite.
